### PR TITLE
OIDC Authentication, and Cloud CLI Contexts

### PR DESCRIPTION
**Pull request by nickmisasi**
#### Summary
This PR introduces 2 major features

##### OIDC Authentication. 
Defaulted to off, unless the server is provided with a Client ID, Secret, and Organization URL. 
Setting the aforementioned parameters will result in the provisioner validating requests to /api/* contain a valid Access Token from Azure AD. 
Authentication for CLI's is handled using Device Authorization grants, via `cloud login` command. Authentication data (access tokens, refresh tokens, etc) are stored in a _context_

Screenshots:
Login page of Azure AD (Browser auto-opens)
image

Logs in terminal when initiated:
image

##### CLI Contexts
Most users have their own custom hacks or bash scripts to automate the passing of `--server` to handle interacting with different environments. This PR introduces the concept of contexts, stored in `~/.cloud/contexts.json`

Contexts store:
- Server URL
- Auth Data (like your access and refresh tokens)
- An alias (for switching)
- Client ID, and Org URL to kick off authorization flow (if configured)

A new CLI option now exists `cloud contexts <command>` with the following options:
```
Manipulate local contexts for the Cloud CLI

Usage:
  cloud contexts [command]

Available Commands:
  create      Create a new context.
  delete      Delete a context.
  get         Get details of a specific context.
  list        List all contexts.
  set-current Set the current context.
  update      Update a context.
```
Additionally, a `cloud login` command has been added, which allows you to retrigger the login flow for a particular context, should you run into authentication issues. 
```
cloud login --help
Login to the Mattermost Cloud provisioning server.

Usage:
  cloud login [flags]

Flags:
      --context string   The name of the context to use.
  -h, --help             help for login
  ```

The `--server` parameter has been removed. Instead, use `cloud context create` to create a context for each Provisioner environment you interact with. To switch between them, use `cloud context set-current`. All `cloud` commands will initiate based off of the currently set context (todo: support `--context` to switch contexts for a single command)

On first run of the Cloud CLI with this change, the contexts.json file will be created for you, containing one context with alias `local` pointing at `http://localhost:8075`. 

When creating a new context, if you provide a Client ID and OrgURL, the CLI will automatically initiate the login flow for you. To skip the authentication flow, pass the argument `--skip-auth`

 

#### Ticket Link
https://mattermost.atlassian.net/browse/CLD-8406
#### Release Note

```release-note
Support for validating OAuth JWT tokens for authentication purposes
Addition of `contexts` command, to enable creating, saving, and switching between different provisioner environments
```

### CODE REVIEW

File: internal/auth/device_authorization.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"os/exec"
+	"runtime"
+	"strings"
+	"time"
+
+	"golang.org/x/oauth2"
```
Comment (nickmisasi) [inline internal/auth/device_authorization.go:15]: Borrowed the OAuth2 built in types for some marshaling, but couldn't use it for token exchange here as MS requires a custom scope to be passed, and this library doesn't support doing that when using refresh tokens

**Review by esarafianou — CHANGES_REQUESTED** [review]
First round of comments, but I'll re-review this.

File: cmd/cloud/main.go [hunk 2]
```diff
@@ -18,11 +21,51 @@ import (
 
 var instanceID string
 
+func isCommandThatSkipsAuth(cmd *cobra.Command) bool {
+	return cmd.Use == "migrate" || cmd.Use == "server" || cmd.Use == "login" || cmd.Parent().Use == "contexts"
```
[the remark this answers is not in the document]
Comment (nickmisasi) [reply cmd/cloud/main.go:25]: It's the database migration command which is run directly against the database

File: cmd/cloud/server_flag.go [hunk 3]
```diff
@@ -158,6 +158,24 @@ type serverFlagChanged struct {
...
+	RestrictedClientAllowedEndpointsList []string
+	JWKSURL                              string
+}
+
+func (flags *serverAuthFlags) addFlags(command *cobra.Command) {
+	command.Flags().StringVar(&flags.Issuer, "issuer", "", "The OAuth2 issuer URL.")
+	command.Flags().StringVar(&flags.Audience, "audience", "", "The OAuth2 audience.")
+	command.Flags().StringVar(&flags.TokenEndpoint, "token-endpoint", "", "The OAuth2 token endpoint.")
+	command.Flags().StringSliceVar(&flags.RestrictedClientIDs, "restricted-client-ids", []string{}, "The list of restricted client IDs.")
+	command.Flags().StringSliceVar(&flags.RestrictedClientAllowedEndpointsList, "restricted-client-allowed-endpoints", []string{}, "The list of restricted endpoints.")
```
[the remark this answers is not in the document]
Comment (nickmisasi) [reply cmd/cloud/server_flag.go:175]: This is run through IaC YAML manifests anywhere that the authentication would matter - see [here](https://git.internal.mattermost.com/cloud-sre/kubernetes-workloads/gitops-platform/-/blob/main/apps/prod/manifests/cnc/provisioner/provisioner_deployment.yaml?ref_type=heads#L34) we currently set allow-list-cidr-ranges and vpn-cidr ranges similarly to the way these would be set here, so I expect we should be ok from the human error side

File: model/client.go [hunk 4]
```diff
@@ -41,6 +48,75 @@ func NewClientWithHeaders(address string, headers map[string]string) *Client {
 	}
 }
 
+func NewClientWithOAuth(address string, headers map[string]string, clientID, clientSecret, tokenEndpoint string) *Client {
```
[the remark this answers is not in the document]
Comment (nickmisasi) [reply model/client.go:51]: This is the client, which is consumed by calling packages like the CWS, Matterwick, Cloud Plugin, etc. This function will be used there

File: internal/auth/device_authorization.go [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+			log.Print(err)
+			return AuthorizationResponse{}, nil
+		}
+		switch token.Error {
+		case "":
+			// Successfully received a token
+			return token, nil
+		case "authorization_pending":
+			// Do nothing, just wait
+		case "access_denied":
```
[the remark this answers is not in the document]
Comment (nickmisasi) [reply internal/auth/device_authorization.go:54]: Left over from Okta. Removed!

File: cmd/cloud/contexts.go [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+				return fmt.Errorf("context '%s' not found", contextName)
+			}
+
+			delete(contexts.Contexts, contextName)
+
+			if len(contexts.Contexts) == 0 {
+				contexts.CurrentContext = ""
+			} else {
+				// Set the Current Context to the first context in the map
+				contexts.CurrentContext = reflect.ValueOf(contexts.Contexts).MapKeys()[0].String()
```
[the remark this answers is not in the document]
Comment (nickmisasi) [reply cmd/cloud/contexts.go:274]: I've adjusted this to only change the CurrentContext in the event that the CurrentContext is what was deleted

File: internal/auth/cli.go [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+	if token.Valid() {
+		return auth, nil
+	}
+
+	newToken, err := Refresh(ctx, config, auth.RefreshToken)
+	if err != nil {
+		return nil, err
+	}
+
+	auth.AccessToken = newToken.AccessToken
```
[the remark this answers is not in the document]
Comment (nickmisasi) [reply internal/auth/cli.go:39]: I had left this intentionally as a "feature" to enforce re-authentications more frequently, but I guess we can defer that functionality to the auth side. I've changed it to match your suggestion but if you'd like me to change it back let me know.

File: internal/auth/device_authorization.go [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+}
+
+func NewOAuth2Config(ctx context.Context, orgURL string, clientID string) (*oauth2.Config, error) {
+	return &oauth2.Config{
+		ClientID: clientID,
+		Endpoint: oauth2.Endpoint{
+			AuthURL:  orgURL + "/oauth2/v2.0/devicecode",
+			TokenURL: orgURL + "/oauth2/v2.0/token",
+		},
+		Scopes: []string{"openid", "profile", "offline_access", "api://provisioner/provisioner"},
```
[the remark this answers is not in the document]
Comment (nickmisasi) [reply internal/auth/device_authorization.go:74]: `openid` scope is required when using OIDC. It's also what returns the `sub` field which is how we identify the user. If you don't pass it, you get an error `The provided value for the input parameter 'scope' is not valid. The scope profile offline_access is not valid. openid scope is required.`

**Comment by nickmisasi** [comment]
Getting rate limited on a build step. Will try again later

File: internal/auth/device_authorization.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (esarafianou) [reply internal/auth/device_authorization.go:74]: That's interesting, thanks for checking!

File: internal/auth/cli.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (esarafianou) [reply internal/auth/cli.go:39]: I see what you mean. Can we actually revert those changes? Apologies for that! 

Reasoning: 
1.  For public clients that are not single-page applications (SPAs), such as those utilizing the device authorization grant (device flow), the default refresh token lifetime is 90 days. This is already a lot for a public client but it's non-configurable. 
2. Each time a refresh token is used to acquire a new access token, Microsoft Entra ID issues a new refresh token, that inherits the same default lifetime as the original. As long as the refresh token is used within its valid period, it is continuously renewed, potentially allowing indefinite access without requiring the user to re-authenticate.

File: model/client.go [hunk 9]
```diff
@@ -41,6 +48,75 @@ func NewClientWithHeaders(address string, headers map[string]string) *Client {
...
+func (c *Client) getToken() (*oauth2.Token, error) {
+	c.tokenMutex.Lock()
+	defer c.tokenMutex.Unlock()
+
+	// Check if we already have a valid token
+	if c.token != nil && c.token.Valid() {
+		return c.token, nil
+	}
+
+	fmt.Println(c.oauthConfig)
```
Comment (esarafianou) [inline model/client.go:109]: This is printing the client secret. We should either print specific fields of the oauthConfig or remove this altogether.

**Review by esarafianou — CHANGES_REQUESTED** [review]
2 last comments, looks great otherwise!

File: internal/auth/cli.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (nickmisasi) [reply internal/auth/cli.go:39]: Reverted back to not using the new RefreshToken

**Review by esarafianou — APPROVED** [review]

File: cmd/cloud/clicontext/clicontext.go [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	contextFilePath := filepath.Join(homeDir, ".cloud", "contexts.json")
+
+	var contextsData Contexts
+	if _, statErr := os.Stat(contextFilePath); errors.Is(statErr, os.ErrNotExist) {
+		contextsData = bootstrapFirstContext()
+		WriteContexts(&contextsData)
```
Comment (gabrieljackson) [inline cmd/cloud/clicontext/clicontext.go:72]: We are ignoring the return error here and this seems to be hiding an issue where bootstrapping will fail on a completely new bootstrap. I believe the issue is the `.cloud` directory isn't being created.

File: cmd/cloud/clicontext/clicontext.go [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+
+}
+
+func ReadContexts() (*Contexts, error) {
+	homeDir, err := os.UserHomeDir()
+	if err != nil {
+		return nil, err
+	}
+
+	contextFilePath := filepath.Join(homeDir, ".cloud", "contexts.json")
```
Comment (gabrieljackson) [inline cmd/cloud/clicontext/clicontext.go:67]: Let's take this logic and put it into a reusable helper function to rule out any typos on the default context file path.

**Review by gabrieljackson — CHANGES_REQUESTED** [review]
Heya, have some initial feedback from my testing.

**Review by gabrieljackson — APPROVED** [review]
Looks great! Let's get this merged so we can start testing it.

**Pull request merged**